crypto/ed25519: Update the godocs

ValarDragon opened the pull request:
This commit updates the godocs for the package, and adds an optimization
to the privkey.Pubkey() method.

The optimization is that in golang, the private key (due to interface
compatibility reasons) has a copy of the public key stored inside of it.
Therefore if this copy has already been computed, there is no need to
recompute it.

<!-- Thanks for filing a PR! Before hitting the button, please check the following items.-->

* [X] Updated all relevant documentation in docs - no API changes
* [X] Updated all code comments where relevant
* [X] Wrote tests - n/a
* [X] Updated CHANGELOG.md - not user facing

melekes commented on crypto/ed25519/ed25519.go:
```diff
@@ -117,16 +151,19 @@ func GenPrivKeyEd25519FromSecret(secret []byte) PrivKeyEd25519 {
 }
 
+// Bytes marshals the PubKey using amino encoding.
+// Amino encoding was chosen due its portability
```
not sure if it's the best place to talk about our reasoning

Liamsi replied:
I agree.

Liamsi commented on crypto/ed25519/ed25519.go:
```diff
@@ -47,18 +47,37 @@ func init() {
 // Implements crypto.PrivKey
 type PrivKeyEd25519 [64]byte
 
+// Bytes marshals the privkey using amino encoding
```
Nitpicking: I think we want full sentences with a full stop at the end.

Liamsi commented on crypto/ed25519/ed25519.go:
```diff
@@ -47,18 +47,37 @@ func init() {
 	return SignatureEd25519(*signatureBytes), nil
 }
 
+// PubKey gets the corresponding public key from the private key
```
Nit: full stop

Liamsi commented on crypto/ed25519/ed25519.go:
```diff
@@ -73,42 +92,57 @@ func (privKey PrivKeyEd25519) Equals(other crypto.PrivKey) bool {
-func (privKey PrivKeyEd25519) ToCurve25519() *[32]byte {
+// ToCurve25519 takes a private key and returns its representation on
+// Curve25519. Curve25519 is birationally equivalent to Edwards25519,
+// which Ed25519 uses internally. This method is for intended use in
```
and is internally used by Ed25519?

Liamsi commented on crypto/ed25519/ed25519.go:
```diff
@@ -73,42 +92,57 @@ func (privKey PrivKeyEd25519) Equals(other crypto.PrivKey) bool {
+// ToCurve25519 takes a private key and returns its representation on
+// Curve25519. Curve25519 is birationally equivalent to Edwards25519,
+// which Ed25519 uses internally. This method is for intended use in
+// Diffie Hellman key exchanges. (i.e. X25519)
```
This method is intended to be used in ...

Liamsi commented on crypto/ed25519/ed25519.go:
```diff
@@ -73,42 +92,57 @@ func (privKey PrivKeyEd25519) Equals(other crypto.PrivKey) bool {
 }
 
-// Deterministically generates new priv-key bytes from key.
+// Generate deterministically generates a new priv-key bytes from key.
```
s/generates/derives (?)

Liamsi commented on crypto/ed25519/ed25519.go:
```diff
@@ -73,42 +92,57 @@ func (privKey PrivKeyEd25519) Equals(other crypto.PrivKey) bool {
+// Generate deterministically generates a new priv-key bytes from key.
+// The privkey is generated as Sha256(amino_encode({privkey, index}))
+// Note that we are using the non-standard golang definition for ed25519
+// privkeys that includes the pubkey appended to it
```
"Note that similar to golang/x/crypto/ed25519 we include the private key ..."
Maybe, even add a link to: https://github.com/golang/crypto/blob/a49355c7e3f8fe157a85be2f77e6e269a0f89602/ed25519/ed25519.go#L48-L52 ?

The way we derive the key here is non-standard anyways?

Liamsi commented on crypto/ed25519/ed25519.go:
```diff
@@ -73,42 +92,57 @@ func (privKey PrivKeyEd25519) Equals(other crypto.PrivKey) bool {
 	newKey := new([64]byte)
 	copy(newKey[:32], newBytes)
+	// Caches the pubkey in the latter 32 bytes of the privkey
+	// (this is a mutative function)
```
does this clarify / help? It does not mutate the underlying `privKey`? (What is a "mutative function" and why is it relevant in that context?) I'd propose to remove this.

ValarDragon replied:
This does mutate the private key object. (I found this quite confusing tbh, it is what caused #1999)

Perhaps changing it to 
" this mutates the last 32 bytes of the privkey" would be more clear?

Liamsi replied:
It is confusing indeed. But wait, it does not mutate the private on which it is called on (which is represented by the var `privKey`, right?), it does indeed "mutate" the freshly created private-key though which is called `newKey`. Not sure if "mutative" makes sense in this context then (because of course the method "mutates" the object which it creates). Or am I getting this wrong?

ValarDragon replied:
`ed25519.MakePublicKey(newKey)` alters the last 32 bytes of `newKey`

ValarDragon commented:
It was the right base, since thats where I refactored into an ed25519 file. I squashed that branch, which is whats causing all of the merge conflicts here. I don't really know how to fix this tho. Should I open a new PR on develop, once `dev/refactor_crypto` is merged?

Liamsi commented:
> Should I open a new PR on develop that cherry-picks these commits, once dev/refactor_crypto is merged?

I think you can either rebase this branch on (origin) dev/refactor_crypto or merge dev/refactor_crypto into this branch. In both cases you'd need to resolve the merge conflicts above.

ValarDragon commented:
I'm amazed this worked, but I ran `git rebase -i HEAD~10 --onto dev/refactor_crypto`, and dropped all commits except the ones I wanted, and squashed those together into one.
